### ISSUE TITLE
Incorrect result when using `map_elements` with a list of list return type 

### ISSUE THREAD

**Issue by heshamdar** [pola-rs/polars#18703]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

def single_list(b):

    if b[0] == 0:
        return []
    
    return [0,1,2]

def double_list(b):

    if b[0] == 0:
        return [[]]
    
    return [[0,1,2]]


test = pl.DataFrame({
                     'b':[[0,0,0],[1, 256, 255]]}, 
                    schema={'b':pl.List(pl.Int64)})

result = test.with_columns(

    pl.col('b').map_elements(
        lambda x: single_list(x), 
        return_dtype=pl.List(pl.Int64), 
        skip_nulls=False).alias('single_list'),

    pl.col('b').map_elements(
        lambda x: single_list(x), 
        return_dtype=pl.List(pl.Int64), 
        skip_nulls=True).alias('single_list_skip_null'),


    pl.col('b').map_elements(
        lambda x: double_list(x), 
        return_dtype=pl.List(pl.List(pl.Int64)), 
        skip_nulls=False).alias('double_list'),

    pl.col('b').map_elements(
        lambda x: double_list(x), 
        return_dtype=pl.List(pl.List(pl.Int64)), 
        skip_nulls=True).alias('double_list_skip_null'),
)
                         
print(result)

```

### Log output

```shell
shape: (2, 5)
┌───────────────┬─────────────┬──────────────────────┬──────────────────────┬──────────────────────┐
│ b             ┆ single_list ┆ single_list_skip_nul ┆ double_list          ┆ double_list_skip_nul │
│ ---           ┆ ---         ┆ l                    ┆ ---                  ┆ l                    │
│ list[i64]     ┆ list[i64]   ┆ ---                  ┆ list[list[i64]]      ┆ ---                  │
│               ┆             ┆ list[i64]            ┆                      ┆ list[list[i64]]      │
╞═══════════════╪═════════════╪══════════════════════╪══════════════════════╪══════════════════════╡
│ [0, 0, 0]     ┆ []          ┆ []                   ┆ [[]]                 ┆ [[]]                 │
│ [1, 256, 255] ┆ [0, 1, 2]   ┆ [0, 1, 2]            ┆ [[null, null, null]] ┆ [[0, 1, 2]]          │
└───────────────┴─────────────┴──────────────────────┴──────────────────────┴──────────────────────┘

```

### Issue description

When a `map_elements` operation returns a list of list and the `skip_nulls` argument is set to `False` the type of the first returned element seems to be used. In the case where this is an empty list, all subsequent values will have null values. This isn't the case if either
- There is only a single level in the list
-`skip_nulls` is set to `True`

The latter is a bit confusing, since neither the input or the output should be null.

The example shows the the behaviour under the settings described above, with the only failing case shown in the column `double_list`

I think it's perhaps similar to this issue related to this fix, but is still an issue for list of lists (and probably deeper nesting) - https://github.com/pola-rs/polars/pull/18567

### Expected behavior

Similar to the single list case, the type for the nested list should be the same as defined in the `return_dtype` argument, and not inferred based on the first result. Also it shouldn't matter whether the `skip_nulls` argument is `True` or `False` (maybe?)

### Installed versions

<details>

```
--------Version info---------
Polars:              1.7.0
Index type:          UInt32
Platform:            macOS-14.2.1-arm64-arm-64bit
Python:              3.10.4 (main, May 28 2022, 12:23:44) [Clang 13.1.6 (clang-1316.0.21.2.5)]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.10.4
fsspec               <not installed>
gevent               <not installed>
great_tables         0.10.0
matplotlib           3.9.1.post1
nest_asyncio         1.6.0
numpy                1.26.4
openpyxl             <not installed>
pandas               2.2.2
pyarrow              16.1.0
pydantic             <not installed>
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by elipinska** [comment]
I've run into similar behaviour with a list of structs: 

### Reproducible example

```python
import polars as pl

original_df = pl.DataFrame(
    [
        {
            "list_of_dicts": [
                {"foo": "bar", "qux": "quz"},
                {"foo": "baz", "qux": "quy"},
            ],
        },
        {
            "list_of_dicts": [
                {"foo": "bar", "qux": "quz"},
                {"foo": "baz", "qux": "quy"},
            ],
        },
        {
            "list_of_dicts": [
                {"foo": "bar", "qux": "quz"},
                {"foo": "baz", "qux": "quy"},
            ],
        },
    ]
)

transformed_df = original_df.with_columns(
    pl.col("list_of_dicts").map_elements(
        lambda structs: structs, return_dtype=pl.List(pl.Struct), skip_nulls=True
    )
)

print(transformed_df)
```

### Log output
```
┌─────────────────┐
│ list_of_dicts   │
│ ---             │
│ list[struct[0]] │
╞═════════════════╡
│ []              │
│ []              │
│ []              │
└─────────────────┘
```

Versions of polars before 1.7.0 returned the list of structs correctly: 

```
┌────────────────────────────────┐
│ list_of_dicts                  │
│ ---                            │
│ list[struct[2]]                │
╞════════════════════════════════╡
│ [{"bar","quz"}, {"baz","quy"}] │
│ [{"bar","quz"}, {"baz","quy"}] │
│ [{"bar","quz"}, {"baz","quy"}] │
└────────────────────────────────┘
```

Similarly to @heshamdar, I've noticed that setting `skip_nulls=False` when running the example with polars >=1.7.0 *does* return the lists of structs I'm expecting (though it's somewhat confusing why).

**Comment by cmdlineluser** [comment]
@elipinska That example also works without the `return_dtype=` 

It seems like `pl.Struct` now ends up "casting" to a struct with 0 fields.

I have asked in some previous issues where it cropped up if `pl.Struct` is supposed to be allowed as a valid dtype - but have not found an answer yet.

`skip_nulls=False` also seems to cause some odd behaviour.

```python
df = pl.DataFrame({
    "list_of_dicts": [
        [ {"foo": "bar", "qux": "quz"} ]
    ]
})
```
```python
df.with_columns(
    pl.col("list_of_dicts").map_elements(
        lambda structs: [print(structs), structs][-1], 
        skip_nulls=True,
        return_dtype=pl.List(pl.Struct)
    )
)

# shape: (1,)
# Series: '' [struct[2]]
# [
# 	{"bar","quz"}
# ]
# 
# shape: (1, 1)
# ┌─────────────────┐
# │ list_of_dicts   │
# │ ---             │
# │ list[struct[0]] │
# ╞═════════════════╡
# │ []              │
# └─────────────────┘
```

Setting it to False appears to lose the Series type being sent to the UDF

```python
df.with_columns(
    pl.col("list_of_dicts").map_elements(
        lambda structs: [print(structs), structs][-1], 
        skip_nulls=False,
        return_dtype=pl.List(pl.Struct)
    )
)

# [{'foo': 'bar', 'qux': 'quz'}] <- ???
#
# shape: (1, 1)
# ┌─────────────────┐
# │ list_of_dicts   │
# │ ---             │
# │ list[struct[2]] │
# ╞═════════════════╡
# │ [{"bar","quz"}] │
# └─────────────────┘
```

And also seemingly bypasses this `pl.List(pl.Struct)` dtype.

**Comment by cmdlineluser** [comment]
I've traced the `double_list` example to here:

https://github.com/pola-rs/polars/blob/71a8b0543362aadb771c2a446ac93756ea36a124/crates/polars-python/src/series/map.rs#L89-L90

```rust
[crates/polars-python/src/series/map.rs:94:17] avs.clone() = [
    List(
        shape: (1,)
        Series: '' [list[null]]
        [
        	[]
        ],
    ),
    List(
        shape: (1,)
        Series: '' [list[i64]]
        [
        	[0, 1, 2]
        ],
    ),
]
```

So it seems `Series::new` takes type of the first element. I'm not sure if a different Series constructor is supposed to be used with `return_dtype` given instead.


It seems 2 completely different codepaths are taken depending on the value of `skip_nulls` - it's a bit hard to follow - but that seems to be why there is such odd behaviour going on when it is set.
